src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -561,106 +528,71 @@ get_relation_constraint_oid_compat(HeapTuple heapTuple)
 
 
 /*
- * HasForeignKeyToReferenceTable function scans the pgConstraint table to
- * fetch all of the constraints on the given relationId and see if at least one
- * of them is a foreign key referencing to a reference table.
+ * HasForeignKeyToReferenceTable function returns true if any of the foreign
+ * key constraints on the relation with relationOid references to a reference
+ * table.
  */
 bool
-HasForeignKeyToReferenceTable(Oid relationId)
+HasForeignKeyToReferenceTable(Oid relationOid)
 {
-	ScanKeyData scanKey[1];
-	int scanKeyCount = 1;
-	bool hasForeignKeyToReferenceTable = false;
+	return list_length(GetForeignKeyOidsToReferenceTables(relationOid)) > 0;
+}
 
-	Relation pgConstraint = heap_open(ConstraintRelationId, AccessShareLock);
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid, BTEqualStrategyNumber, F_OIDEQ,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint,
-													ConstraintRelidTypidNameIndexId,
-													true, NULL,
-													scanKeyCount, scanKey);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+/*
+ * GetForeignKeyOidsToReferenceTables function returns list of OIDs for the
+ * foreign key constraints on the given relationOid that are referencing to
+ * reference tables.
+ */
+static List *
+GetForeignKeyOidsToReferenceTables(Oid relationOid)
+{
+	bool excludeSelfReference = false;
+	List *foreignKeyOids = GetTableForeignConstraintOidsInternal(relationOid,
```

Comment:
I find passing `Anum_pg_constraint_conrelid ` (and  also `excludeSelfReference ` ) is a little confusing. Can't we define an enum `ENUM` like
```
FETCH_REFERENCED_FOREIGN_CONSTRAINTS,
FETCH_REFERENCING_FOREIGN_CONSTRAINTS,
FETCH_SELF_REFERENCE_FOREIGN_CONSTRAINTS
```

And, the caller can pass any combination of the enum  like  `FETCH_REFERENCED_FOREIGN_CONSTRAINT | FETCH_REFERENCING_FOREIGN_CONSTRAINT`, and the internal function behaves accordingly

Reply:
(similar for `GetTableForeignConstraintCommandsInternal `)

And,  please pick a better enum name convention, I just quickly made up

Reply:
That would also reduce the usage of an additional parameter `excludeSelfReference`, I will apply this too

Reply:
```c
/* 
 * Flags that can be passed to GetTableForeignConstraintOidsInternal to indicate
 * which foreign key constraint OIDs are extracted
 */
#define EXTRACT_REFERENCING_CONSTRAINTS   0x0001
#define EXTRACT_REFERENCED_CONSTRAINTS    0x0010
#define EXCLUDE_SELF_REFERENCES		  0x0100
```

What about those above ?